adapter/kubernetes/cache.go

```diff
@@ -0,0 +1,193 @@
+		pods          cache.SharedInformer
+		mutationsChan chan resourceMutation
+	}
+
+	// used to send updates to the logger
+	resourceMutation struct {
+		kind eventType
+		obj  interface{}
+	}
+
+	eventType int
+)
+
+const (
+	addition eventType = iota
+	update
+	deletion
+)
+
+// mutationBufferSize sets the limit on how many mutation events can be
+// outstanding at any moment. 100 is chosen as a reasonable default to start.
+// TODO: make this configurable
+const mutationBufferSize = 100
+
+// errorDelay controls how long the logger waits when encountering an error
+// during logging. This should only ever happen when the underlying cache has
+// not yet synced (at which point we need to wait before doing any further
+// processing).
+// TODO: make this configurable
+const errorDelay = 1 * time.Second
+
+// Responsible for setting up the cacheController, based on the supplied client.
+// It configures the index informer to list/watch pods and send update events
+// to a mutations channel for processing (in this case, logging).
+func newCacheController(clientset *kubernetes.Clientset, namespace string, refreshDuration time.Duration, logger adapter.Logger) cacheController {
+	c := &controllerImpl{
+		clientset:     clientset,
+		logger:        logger,
+		mutationsChan: make(chan resourceMutation, mutationBufferSize),
+	}
+
+	c.pods = cache.NewSharedIndexInformer(
+		&cache.ListWatch{
+			ListFunc: func(opts v1.ListOptions) (runtime.Object, error) {
+				return clientset.Pods(namespace).List(opts)
+			},
+			WatchFunc: func(opts v1.ListOptions) (watch.Interface, error) {
+				return clientset.Pods(namespace).Watch(opts)
+			},
+		},
+		&v1.Pod{},
+		refreshDuration,
+		cache.Indexers{},
+	)
+
+	// log all events, for debug
+	eventErr := c.pods.AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.mutationsChan <- resourceMutation{addition, obj}
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.mutationsChan <- resourceMutation{deletion, obj}
+			},
+			UpdateFunc: func(old, cur interface{}) {
+				if !reflect.DeepEqual(old, cur) {
+					c.mutationsChan <- resourceMutation{update, cur}
+				}
+			},
+		},
+	)
+
+	if eventErr != nil {
+		c.logger.Warningf("could not add logging event handlers: %v", eventErr)
+	}
+
+	return c
+}
+
+// Run starts the logger and the controller for the pod cache.
+func (c *controllerImpl) Run(stop <-chan struct{}) {
+	go c.runLogger(stop)
+	go c.pods.Run(stop)
+	c.logger.Infof("cluster cache started")
+	<-stop
+	c.logger.Infof("cluster cache updating terminated")
+}
+
+// runLogger is responsible for pulling event updates off of the mutations
+// channel and logging them via the configured logger.
+func (c *controllerImpl) runLogger(stop <-chan struct{}) {
+	for {
+		select {
+		case mutation := <-c.mutationsChan:
+			err := c.log(mutation.obj, mutation.kind)
+			if err != nil {
+				c.logger.Infof("event logging failed, will retry: %v", err)
+				time.Sleep(errorDelay)
```

Comment:
Does it matter that this delays runLogger from stopping? Does this need to be replaced by select over `time.After(errorDelay)` and `<-stop`?